packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -1221,6 +1230,42 @@ function updateTransition(
   return [start, isPending];
 }
 
+let clientId: number = 0;
+
+function mountOpaqueReference(): string | IdObject {
+  const [id, setId] = mountState(() => {
+    if (getIsHydrating()) {
+      return {
+        $$typeof: REACT_OPAQUE_OBJECT_TYPE,
+        _setId: () => {
+          setId(() => 'c_' + (clientId++).toString());
+        },
+        toString() {
+          invariant(
+            false,
+            'The object passed back from useOpaqueReference is meant to be passed through to ' +
+              'attributes only. Do not read the value directly.',
+          );
+        },
+        valueOf() {
+          invariant(
+            false,
+            'The object passed back from useOpaqueReference is meant to be passed through to ' +
+              'attributes only. Do not read the value directly.',
+          );
+        },
+      };
+    } else {
+      return 'c_' + (clientId++).toString();
```

Comment:
This should also move to `ReactDOM` since it's specific to how an "Opaque Reference" is constructed in the DOM.

Also, don't call `toString()`. Rely on `+ ''` casting instead like we do elsewhere. `.toString()` becomes a dynamic look up which can be monkey patched by `Number.prototype`. It's more bytes and less efficient.